File: murmer_server/src/permissions.rs
//! Server-wide permission bitmask shared by every authorization check.
//!
//! A role grants a set of capabilities encoded as bit flags. A user's
//! effective permissions are the union (bitwise OR) of the built-in
//! `@everyone` baseline role and every role assigned to them; holding
//! [`ADMINISTRATOR`] short-circuits every check. The mask is mirrored on the
//! client in `murmer_client/src/lib/chat/permissions.ts` — keep the two in
//! sync when adding or renaming a flag.
//!
//! Enforcement is always server-side: clients use the same flags only to show
//! or hide UI, never as the source of truth.

/// Bitmask type used for role permissions throughout the server.
pub type Permissions = u64;

/// See channels and read message history. Newly enforced by the role system.
pub const VIEW_CHANNELS: Permissions = 1 << 0;
/// Send chat messages and add reactions. Newly enforced by the role system.
pub const SEND_MESSAGES: Permissions = 1 << 1;
/// Delete or pin messages authored by other users.
pub const MANAGE_MESSAGES: Permissions = 1 << 2;
/// Create, delete, move, reorder and re-topic channels and categories.
pub const MANAGE_CHANNELS: Permissions = 1 << 3;
/// Create, edit, rename and delete wiki pages.
pub const MANAGE_WIKI: Permissions = 1 << 4;
/// Create, edit, delete, reorder and assign roles.
pub const MANAGE_ROLES: Permissions = 1 << 5;
/// Add or remove custom server emojis.
pub const MANAGE_EMOJIS: Permissions = 1 << 6;
/// Edit server identity, stat tracking, screen-share cap and other settings.
pub const MANAGE_SERVER: Permissions = 1 << 7;
/// Query server details such as the running version.
pub const VIEW_SERVER_INFO: Permissions = 1 << 8;
/// View other users' self-reported connection stats.
pub const VIEW_CONNECTION_STATS: Permissions = 1 << 9;
/// Kick members from the server.
pub const KICK_MEMBERS: Permissions = 1 << 10;
/// Ban members from the server.
pub const BAN_MEMBERS: Permissions = 1 << 11;
/// Mute members.
pub const MUTE_MEMBERS: Permissions = 1 << 12;
/// Grants every permission and bypasses hierarchy checks (the Owner role).
pub const ADMINISTRATOR: Permissions = 1 << 13;

/// Union of every defined permission flag. Used to reject unknown bits from
/// clients and to expand [`ADMINISTRATOR`] into a concrete mask.
pub const ALL: Permissions = VIEW_CHANNELS
    | SEND_MESSAGES
    | MANAGE_MESSAGES
    | MANAGE_CHANNELS
    | MANAGE_WIKI
    | MANAGE_ROLES
    | MANAGE_EMOJIS
    | MANAGE_SERVER
    | VIEW_SERVER_INFO
    | VIEW_CONNECTION_STATS
    | KICK_MEMBERS
    | BAN_MEMBERS
    | MUTE_MEMBERS
    | ADMINISTRATOR;

/// Baseline permissions granted to every user through the `@everyone` role.
/// Keeps a fresh or unadministered server usable: everyone can read and chat.
pub const DEFAULT_EVERYONE: Permissions = VIEW_CHANNELS | SEND_MESSAGES;

/// Default permissions seeded for the built-in `Mod` role. Mirrors the legacy
/// Mod capabilities: manage channels/wiki/emojis, moderate messages and act
/// against lower-ranked members.
pub const DEFAULT_MOD: Permissions = DEFAULT_EVERYONE
    | MANAGE_MESSAGES
    | MANAGE_CHANNELS
    | MANAGE_WIKI
    | MANAGE_EMOJIS
    | KICK_MEMBERS
    | BAN_MEMBERS
    | MUTE_MEMBERS;

/// Default permissions seeded for the built-in `Admin` role: everything a Mod
/// can do plus server settings and read-only server/connection insight.
pub const DEFAULT_ADMIN: Permissions =
    DEFAULT_MOD | MANAGE_SERVER | VIEW_SERVER_INFO | VIEW_CONNECTION_STATS;

/// Default permissions seeded for the built-in `Owner` role.
pub const DEFAULT_OWNER: Permissions = ADMINISTRATOR;

/// Whether a permission mask satisfies `required`. [`ADMINISTRATOR`] grants
/// everything.
pub fn mask_allows(mask: Permissions, required: Permissions) -> bool {
    mask & ADMINISTRATOR != 0 || mask & required == required
}

/// Whether a mask only uses defined flags (rejects unknown bits from clients).
pub fn is_valid_mask(mask: Permissions) -> bool {
    mask & !ALL == 0
}

File: murmer_client/src/lib/chat/permissions.ts
/**
 * Client mirror of the server permission bitmask
 * (`murmer_server/src/permissions.rs`). Keep the flag values in sync.
 *
 * Every gate here is cosmetic: it only decides whether to show or enable UI.
 * The server re-checks every action, so a tampered client gains nothing.
 */
import type { RoleDef } from '../types';

export const PERMISSIONS = {
  VIEW_CHANNELS: 1 << 0,
  SEND_MESSAGES: 1 << 1,
  MANAGE_MESSAGES: 1 << 2,
  MANAGE_CHANNELS: 1 << 3,
  MANAGE_WIKI: 1 << 4,
  MANAGE_ROLES: 1 << 5,
  MANAGE_EMOJIS: 1 << 6,
  MANAGE_SERVER: 1 << 7,
  VIEW_SERVER_INFO: 1 << 8,
  VIEW_CONNECTION_STATS: 1 << 9,
  KICK_MEMBERS: 1 << 10,
  BAN_MEMBERS: 1 << 11,
  MUTE_MEMBERS: 1 << 12,
  ADMINISTRATOR: 1 << 13
} as const;

export type PermissionKey = keyof typeof PERMISSIONS;

/** Union of every defined permission flag. */
export const ALL_PERMISSIONS = Object.values(PERMISSIONS).reduce((acc, bit) => acc | bit, 0);

/** Whether a permission mask satisfies `flag`. Administrator grants everything. */
export function hasPermission(mask: number, flag: number): boolean {
  return (mask & PERMISSIONS.ADMINISTRATOR) !== 0 || (mask & flag) === flag;
}

/**
 * The effective permission mask for a set of assigned role ids: the union of
 * the default `@everyone` role and every assigned role. Administrator expands
 * to the full mask.
 */
export function computeMask(defs: RoleDef[], assignedIds: number[]): number {
  let mask = defs.find((d) => d.isDefault)?.permissions ?? 0;
  const byId = new Map(defs.map((d) => [d.id, d]));
  for (const id of assignedIds) {
    const def = byId.get(id);
    if (def) mask |= def.permissions;
  }
  return (mask & PERMISSIONS.ADMINISTRATOR) !== 0 ? ALL_PERMISSIONS : mask;
}

/**
 * The highest hierarchy position across a user's roles (the default role's
 * position is the floor). Administrators sit above everyone
 * (`Number.POSITIVE_INFINITY`).
 */
export function computeTopPosition(defs: RoleDef[], assignedIds: number[]): number {
  const def = defs.find((d) => d.isDefault);
  let position = def?.position ?? 0;
  let admin = def ? (def.permissions & PERMISSIONS.ADMINISTRATOR) !== 0 : false;
  const byId = new Map(defs.map((d) => [d.id, d]));
  for (const id of assignedIds) {
    const role = byId.get(id);
    if (!role) continue;
    position = Math.max(position, role.position);
    if ((role.permissions & PERMISSIONS.ADMINISTRATOR) !== 0) admin = true;
  }
  return admin ? Number.POSITIVE_INFINITY : position;
}

export interface PermissionMeta {
  key: PermissionKey;
  flag: number;
  label: string;
  description: string;
}

export interface PermissionGroup {
  title: string;
  permissions: PermissionMeta[];
}

/** Grouping used by the dashboard's role editor checkboxes. */
export const PERMISSION_GROUPS: PermissionGroup[] = [
  {
    title: 'General',
    permissions: [
      {
        key: 'VIEW_CHANNELS',
        flag: PERMISSIONS.VIEW_CHANNELS,
        label: 'View channels',
        description: 'See channels and read message history.'
      },
      {
        key: 'SEND_MESSAGES',
        flag: PERMISSIONS.SEND_MESSAGES,
        label: 'Send messages',
        description: 'Post messages and add reactions.'
      }
    ]
  },
  {
    title: 'Messages',
    permissions: [
      {
        key: 'MANAGE_MESSAGES',
        flag: PERMISSIONS.MANAGE_MESSAGES,
        label: 'Manage messages',
        description: "Delete and pin other members' messages."
      }
    ]
  },
  {
    title: 'Members',
    permissions: [
      {
        key: 'KICK_MEMBERS',
        flag: PERMISSIONS.KICK_MEMBERS,
        label: 'Kick members',
        description: 'Disconnect members from the server.'
      },
      {
        key: 'BAN_MEMBERS',
        flag: PERMISSIONS.BAN_MEMBERS,
        label: 'Ban members',
        description: 'Ban and unban members.'
      },
      {
        key: 'MUTE_MEMBERS',
        flag: PERMISSIONS.MUTE_MEMBERS,
        label: 'Mute members',
        description: 'Mute and unmute members.'
      }
    ]
  },
  {
    title: 'Management',
    permissions: [
      {
        key: 'MANAGE_CHANNELS',
        flag: PERMISSIONS.MANAGE_CHANNELS,
        label: 'Manage channels',
        description: 'Create, edit, reorder and delete channels and categories.'
      },
      {
        key: 'MANAGE_WIKI',
        flag: PERMISSIONS.MANAGE_WIKI,
        label: 'Manage wiki',
        description: 'Create, edit and delete wiki pages.'
      },
      {
        key: 'MANAGE_EMOJIS',
        flag: PERMISSIONS.MANAGE_EMOJIS,
        label: 'Manage emojis',
        description: 'Add and remove custom server emojis.'
      },
      {
        key: 'MANAGE_ROLES',
        flag: PERMISSIONS.MANAGE_ROLES,
        label: 'Manage roles',
        description: 'Create, edit, delete and assign roles below your own.'
      },
      {
        key: 'MANAGE_SERVER',
        flag: PERMISSIONS.MANAGE_SERVER,
        label: 'Manage server',
        description: 'Edit server identity, stats, screen-share and other settings.'
      },
      {
        key: 'VIEW_SERVER_INFO',
        flag: PERMISSIONS.VIEW_SERVER_INFO,
        label: 'View server info',
        description: 'See server details such as the running version.'
      },
      {
        key: 'VIEW_CONNECTION_STATS',
        flag: PERMISSIONS.VIEW_CONNECTION_STATS,
        label: 'View connection stats',
        description: "View other members' connection quality."
      },
      {
        key: 'ADMINISTRATOR',
        flag: PERMISSIONS.ADMINISTRATOR,
        label: 'Administrator',
        description: 'Grants every permission and bypasses the hierarchy. Use sparingly.'
      }
    ]
  }
];
